### PR TITLE
Create object

### PR DESCRIPTION
**Pull request by unicornzero**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -12,5 +12,6 @@ end
 
 group :test do
   gem 'rspec'
+  gem 'byebug'
```
Comment (garston) [inline Gemfile]: I'm assuming we mean to check this in?

File: lib/wsapi/session.rb [hunk 2]
```diff
@@ -161,6 +169,20 @@ def wsapi_get(url, opts = {})
       response
     end
 
+    def wsapi_post(url, opts = {})
```
Comment (garston) [inline lib/wsapi/session.rb]: 🌶 alphabetize

File: lib/wsapi/session.rb [hunk 3]
```diff
@@ -161,6 +169,20 @@ def wsapi_get(url, opts = {})
...
+      response = wsapi_request_with_refresh_token(:post, url, opts)
+      check_response_for_errors!(response)
+
+      response
+    end
+
+    def wsapi_put(url, opts = {})
```
Comment (garston) [inline lib/wsapi/session.rb]: Refactor/combine with `wsapi_post` since only diff is `:post` vs `:put`?

File: Gemfile (hunk 1 above)
Comment (unicornzero) [inline Gemfile]: yep, so I don't have to re-add it every time

**Comment by unicornzero** [comment]
bump

File: lib/wsapi/session.rb [hunk 4]
```diff
@@ -44,20 +44,37 @@ def initialize(session_id, opts = {})
...
 
-    def setup_refresh_token(client_id, client_secret, refresh_token, &block)
-      @oauth2 = {
-        client_id: client_id,
-        client_secret: client_secret,
-        refresh_token: refresh_token,
-        refresh_token_updated: block
-      }
+    def build_object_url(type, id)
+      wsapi_resource_url("#{type}/#{id}").gsub('userstory', 'hierarchicalrequirement')
```
Comment (garston) [inline lib/wsapi/session.rb]: 🌶

File: README.md [hunk 5]
```diff
@@ -90,9 +90,21 @@ get_team_members(project_id, opts = {})
...
+update(object_type, object_id, parameters)
+```
+
+### Create an object with a fields hash
+```
+create(object_type, object_fields)
+```
+
+### Build a wsapi object url
+This URL could be used when associating artifacts in the #create or #update methods
```
Comment (garston) [inline README.md]: 🌶

File: lib/wsapi/session.rb [hunk 6]
```diff
@@ -44,20 +44,37 @@ def initialize(session_id, opts = {})
...
 
-    def get_user_subscription
-      response = wsapi_get(wsapi_resource_url("Subscription"))
+    def create(type, fields)
+      response = wsapi_request(:put, wsapi_resource_url("#{type}/create"), {type => fields})
+      Mapper.get_object(response)
+    end
+
+    def get_current_user
+      response = wsapi_get(wsapi_resource_url("User"))
```
Comment (garston) [inline lib/wsapi/session.rb]: Inline?????????

File: lib/wsapi/session.rb [hunk 7]
```diff
@@ -44,20 +44,37 @@ def initialize(session_id, opts = {})
...
+    end
+
+    def get_editors(project_id, opts = {})
+      fetch_with_pages(opts) do |page_query|
+        wsapi_get(wsapi_resource_url("Project/#{project_id}/Editors"), opts.merge(page_query))
+      end
+    end
+
+    def get_project(id)
+      response = wsapi_get(wsapi_resource_url("Project/#{id}"))
```
Comment (garston) [inline lib/wsapi/session.rb]: Inline?????????  Or don't, it's fine.

File: lib/wsapi/session.rb (hunk 6 above)
Comment (unicornzero) [inline lib/wsapi/session.rb]: I like being consistent with the pattern in the other methods. Makes it easier to read.

**Review by unicornzero — COMMENTED** [review]
testing the review changes functionality. This is the summary field.

File: lib/wsapi/session.rb [hunk 8]
```diff
@@ -44,20 +44,37 @@ def initialize(session_id, opts = {})
...
-        refresh_token_updated: block
-      }
+    def build_object_url(type, id)
+      wsapi_resource_url("#{type}/#{id}").gsub('userstory', 'hierarchicalrequirement')
     end
 
-    def get_user_subscription
-      response = wsapi_get(wsapi_resource_url("Subscription"))
+    def create(type, fields)
+      response = wsapi_request(:put, wsapi_resource_url("#{type}/create"), {type => fields})
```
Comment (garston) [inline lib/wsapi/session.rb]: Probably misreading but should this be `{"#{type}" => fields}` as done in `update`?

File: lib/wsapi/session.rb (hunk 7 above)
Comment (unicornzero) [inline lib/wsapi/session.rb]: I like being consistent with the pattern in the other methods. Makes it easier to read.

File: lib/wsapi/session.rb (hunk 8 above)
Comment (unicornzero) [inline lib/wsapi/session.rb]: I think they're basically doing the same thing. Good eye. Not sure if there's a reason to change for "consistency" or not.

Comment (unicornzero) [inline lib/wsapi/session.rb]: we're dealing with symbol to json translation happening as requests are sent, I believe

Comment (unicornzero) [inline lib/wsapi/session.rb]: I'll play with it in IRB for a minute...

Comment (garston) [inline lib/wsapi/session.rb]: They are the same, whatever you feel like is fine.

**Comment by garston** [comment]
:shipit:

File: lib/wsapi/session.rb (hunk 8 above)
Comment (garston) [inline lib/wsapi/session.rb]:
```
irb(main):007:0> k = "story"
=> "story"
irb(main):008:0> {k => 1}
=> {"story"=>1}
irb(main):009:0> {"#{k}" => 2}
=> {"story"=>2}
```

Comment (garston) [inline lib/wsapi/session.rb]:
```
irb(main):010:0> k = :story
=> :story
irb(main):011:0> {k => 1}
=> {:story=>1}
irb(main):012:0> {"#{k}" => 2}
=> {"story"=>2}
```

**Comment by unicornzero** [comment]
bump

**Pull request merged**